Add support for NoiseBench

elenamer opened the pull request:
This PR adds support for [NoiseBench](https://arxiv.org/abs/2405.07609), a subset of CoNLL-03 for English with 7 different label sets, including 6 different types of label noise.

Example usage (intialize dataset with a crowdsourced version of the NER labels):
```
corpus = NER_NOISEBENCH(noise="crowd")
```

Potential issues:

To create the dataset, we need the [CleanCoNLL](https://aclanthology.org/2023.emnlp-main.533.pdf) labels as gold labels. This required running an external script in this implementation: https://github.com/flairNLP/CleanCoNLL/blob/main/create_cleanconll_from_conll03.sh

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+            **corpusargs: The arguments propagated to :meth:'flair.datasets.ColumnCorpus.__init__'.
+        """
+        if noise not in ["clean", "crowd", "crowdbest", "expert", "distant", "weak", "llm"]:
+            raise Exception("Please choose a valid version")
```
It is recommended to raise [ValueError](https://docs.python.org/3/library/exceptions.html#:~:text=Passing%20arguments%20of%20the%20wrong%20type%20(e.g.%20passing%20a%20list%20when%20an%20int%20is%20expected)%20should%20result%20in%20a%20TypeError%2C%20but%20passing%20arguments%20with%20the%20wrong%20value%20(e.g.%20a%20number%20outside%20expected%20boundaries)%20should%20result%20in%20a%20ValueError.) for wrong and unexpected values passed to the argument.

Also, an error message would be more helpful if it listed valid values. Something like:
```python
if noise not in VALID_NOISE_VALUES:
    raise ValueError(f"Unsupported value for noise argument. Got {noise}, expected one of {VALUE_NOISE_VALUES}!")
```

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+        if noise not in ["clean", "crowd", "crowdbest", "expert", "distant", "weak", "llm"]:
+            raise Exception("Please choose a valid version")
+
+        self._set_path(base_path)
```
I would remove this function and just set the `base_path` here in `__init__`:
```python
if base_path:
    self.base_path = Path(base_path)
else:
    self.base_path = flair.cache_root / "datasets" / "noisebench"
```

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+        self.base_path = flair.cache_root / "datasets" / "noisebench" if not base_path else Path(base_path)
+
+    @staticmethod
+    def read_column_file(filename):
```
You are adding a new public method to `NER_NOISEBENCH`. I would avoid this. Also, you are relying on this method only in non-public methods in the class (except in `generate_data_files`, which I would also put as non-public.)

My suggestion is to rename the method to `_read_column_file` and add type annotations for `filename` argument.

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+        return all_x
+
+    @staticmethod
+    def save_to_column_file(filename, list):
```
1. Rename to `_save_to_column_file`
2. Add type annotations to arguments
3. Avoid using the name of built-in modules/types/etc. for argument and variable names. Here, you are using `list` as the name for the arguments. Seems like a more appropriate name is `sentences` which is of type `list`.

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+
+    def generate_data_files(self, filename):
+
+        with open(os.path.join(self.base_path, "annotations_only", "index.txt")) as index_file:
```
You can avoid this `os.path.join` since `base_path` is of type Path. Just use `/` to concatenate the path. Also, we should avoid relying on implicit default values, it is better to explicitly mention that we want to read the file with utf-8:
```python
with open(self.base_path / "annotations_only" / "index.txt", "r", encoding="utf-8") as fp:
    # ...
```

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+        with open(os.path.join(self.base_path, "annotations_only", "index.txt")) as index_file:
+            token_indices = index_file.readlines()
+
+            all_clean_sentences = self.read_column_file(os.path.join(self.cleanconll_base_path, "cleanconll.train"))
```
Same here, use `self.cleanconll_base_path / "cleanconll.train"` instead of `os.path.join`.

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+            )
+
+        # copy test set
+        all_clean_test_sentences = self.read_column_file(os.path.join(self.cleanconll_base_path, "cleanconll.test"))
```
Same here, use `self.cleanconll_base_path / "cleanconll.test"` instead of `os.path.join`.

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+                new_s.append([token[0], token[4]])
+            test_sentences.append(new_s)
+
+        self.save_to_column_file(os.path.join(self.base_path, "clean.test"), test_sentences)
```
Same here, use `self.base_path / "clean.test"` instead of `os.path.join`.

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+            new_s = []
+            for token in s:
+                new_s.append([token[0], token[4]])
+            test_sentences.append(new_s)
```
This can be refactored to a more idiomatic approach with list comprehensions:
```python
for sentence in all_clean_test_sentences:
    new_sentence = [[tokens[0], tokens[4]] for tokens in sentence]
    test_sentences.append(new_sentence)
```

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+            for token, label in zip(clean_sentence, sentence):
+                label[0] = token[0]  # token[0] -> text, token[1] -> BIO label
+        if self.SAVE_TRAINDEV_FILE:
+            self.save_to_column_file(os.path.join(self.base_path, f"{corpus}.traindev"), noisy_labels)
```
Same here, use `self.base_path / f"{corpus}.traindev"` instead of `os.path.join`.

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+
+        noisy_labels = self.read_column_file(os.path.join(self.base_path, "annotations_only", f"{corpus}.traindev"))
+        # print(noisy_labels)
+        # print(token_indices)
```
Remove these debugging prints.

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+    def _merge_tokens_labels(self, corpus, all_clean_sentences, token_indices):
+        # generate NoiseBench dataset variants, given CleanCoNLL, noisy label files and index file
+
+        noisy_labels = self.read_column_file(os.path.join(self.base_path, "annotations_only", f"{corpus}.traindev"))
```
Same here, use `self.base_path / "annotations_only" /  f"{corpus}.traindev"` instead of `os.path.join`.

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+            point = []
+            for line in lines:
+                if "\t" in line.strip():
+                    stripped_line = line.strip().split("\t") if "\t" in line.strip() else line.strip().split(" ")
```
Looks like `stripped_line = line.strip().split("\t") if "\t" in line.strip()` will always be set due to the condition `if "\t" in line.strip()` in line 5076. Refactor to:
```python
if "\t" in line.strip():
    stripped_line = line.strip().split("\t")
else:
    stripped_line = line.strip().split(" ")

fkdosilovic commented on flair/datasets/sequence_labeling.py:
```diff
@@ -4973,6 +4973,209 @@ def _write_instances(cls, version, base_path, split, data):
+    def save_to_column_file(filename, list):
+        with open(filename, "w") as f:
+            for sentence in list:
+                for token in sentence:
```
Rename the iteration variable to `tokens` to convert the meaning that this is a container of elements and not a single token.

alanakbik commented:
@fkdosilovic thanks a lot for your review, this is very helpful! Please note that this is still a draft PR and one major change is still planned.

elenamer commented:
@fkdosilovic Thanks a lot for the review, it was very helpful and I incorporated most of your feedback! 

Also, the CLEANCONLL Corpus is now a part of Flair, so I modified this PR to use it, instead of downloading and running an external .sh script.

alanakbik commented:
@elenamer thanks for adding this!
